### ISSUE TITLE
Enhance stats through SPI

### ISSUE THREAD

**Issue by infearATbaidu** [prestosql/presto#480]
AFAIK Presto only supports some kind of statistics like processRows、userTimeMillis and so on. Is there exists some solution to add user-defined stats to the query results. For example, developers may add custom stats through spi framework, or is there a plan to do such work like this in the future?

Thanks.

**Comment by findepi** [comment]
@infearATbaidu can you describe more the use-case you're trying to solve?

**Comment by infearATbaidu** [comment]
> @infearATbaidu can you describe more the use-case you're trying to solve?

@findepi Thanks for your quick reply。
Consider such condition:
We do some work like user-defined stats in out RecordCursor implementation under SPI framework. Now we want to add these stats to the query results, for example, add to statementStats. Then client can get such stats to do other jobs.

**Comment by findepi** [comment]
@infearATbaidu 
do you mean something like additional fields in `OperatorStats` (and then in `DriverStats`, `PipelineStats`, `TaskStats`, `QueryStats`)?

- what kind of information you want to pass? a number? a text? or a whole lot of things at the same time?
   perhaps you can give a concrete example?
- how this information would be aggregated when going from `OperatorStats` all the way to `QueryStats`?
  see for example `io.prestosql.operator.TaskContext#getTaskStats` -- most information is numbers and the numbers are summed, but some information is max'd
- how would you want to consume this information? in an event listener via `QueryStatistics`?

**Comment by martint** [comment]
@findepi, interestingly, this is something I’ve thought about adding a long time ago. The idea would be to make the set of execution stats the engine can track extensible by allowing connectors to declare additional ones.

We’d need to identify what’s the “essence” of that kind of entity, like:
* what are the possible types (counters, times, etc)
* how do values get combined
* how they get seriialized
* how the are rendered

**Comment by infearATbaidu** [comment]
@findepi 
Yes. We want to add some additional stats in OperatorStats. At the present stage, we only have some simple stats in number format, which meaning is like scan counts in our data source. So we try to enhance RecordCusor under SPI (E.g. add extra method like 'getScanCnts'). Then record these stats in TableScanOperator.
Since these stats can only be generated from TableScanOperator, we can simply merge the stats from all TableScanOperator through the field "operatorSummaries" in the queryStats.

@martint 
--"The idea would be to make the set of execution stats the engine can track extensible by allowing connectors to declare additional ones"
I Couldn’t agree more about this idea!
Since our cases are comparatively simple currently,  we may have some kind of temporary solutions like I said above. But as the cases become more complicated, it seems that some abstractions as you mentioned above are indispensable.

**Comment by raunaqmorarka** [comment]
Connectors can define custom statistics through `io.trino.spi.connector.ConnectorPageSource#getMetrics`
